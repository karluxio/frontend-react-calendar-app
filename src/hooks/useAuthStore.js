import { useDispatch, useSelector } from "react-redux"
import calendarApi from "../api/calendarApi"
import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store"
import { useNavigate } from "react-router-dom"

export const useAuthStore = () => {

  const dispatch = useDispatch()
  const { status, user, errorMessage } = useSelector(state => state.auth)

  const navigate = useNavigate()
  const startLogin = async ({ email, password }) => {

    dispatch(onChecking())

    try {

      const { data } = await calendarApi.post('/auth', { email, password })
      const { token, uid, name } = data
      localStorage.setItem('token', token)
      localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ uid, name }))

    } catch (error) {
      dispatch(onLogout('invalid credentials'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10)
    }
  }

  const startRegister = async ({ email, password, name }) => {

    dispatch(onChecking())

    try {

      const { data } = await calendarApi.post('/auth/new', { email, password, name })
      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime())
      console.log({ uid: data.uid, name: data.name });
      dispatch(onLogin({ uid: data.uid, name: data.name }))

    } catch (error) {
      dispatch(onLogout(error.response.data?.msg || '--'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10);
    }
  }

  const checkAuthToken = async () => {
    const token = localStorage.getItem('token')
    if (!token) return dispatch(onLogout())

    try {

      const { uid, name, token } = await calendarApi.get('/auth/renew')
      localStorage.setItem('token', token)
      localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ uid, name }))

    } catch (error) {
      localStorage.clear()
      console.log({ error })
      dispatch(onLogout())
    }
  }

  const startLogout = () => {
    localStorage.clear()
    dispatch(onLogout())
    dispatch(clearErrorMessage())
    navigate('/auth/login')
  }


  return {
    //* Properties
    errorMessage,
    status,
    user,

    //* Methods
    checkAuthToken,
    startLogin,
    startLogout,
    startRegister
  }
}
