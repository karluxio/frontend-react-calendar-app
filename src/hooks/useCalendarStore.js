import { useDispatch, useSelector } from "react-redux"
import { onAddNewEvent, onSetActiveEvent, onUpdateEvent } from "../store"

export const useCalendarStore = () => {

  const { events, activeEvent } = useSelector(state => state.calendar)
  const dispatch = useDispatch()

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent))
  }

  const startSavingEvent = (calendarEvent) => {
    if (calendarEvent._id) {
      dispatch(onUpdateEvent({ ...calendarEvent }))
    } else {
      dispatch(onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }))
    }
  }


  return {
    //* Properties
    activeEvent,
    events,

    //* Methods
    setActiveEvent,
    startSavingEvent
  }
}