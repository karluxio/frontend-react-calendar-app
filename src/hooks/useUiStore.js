import { useDispatch, useSelector } from "react-redux";
import { onCloseDateModal, onOpenDateModal } from "../store";

export const useUiStore = () => {
  const { isDateModalOpen } = useSelector((state) => state.ui);
  const dispatch = useDispatch();

  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  };

  const toggleDateModal = () => {
    isDateModalOpen ? closeDateModal() : openDateModal();
  };

  return {
    // properties
    isDateModalOpen,

    // functions
    openDateModal,
    closeDateModal,
    toggleDateModal,
  };
};
