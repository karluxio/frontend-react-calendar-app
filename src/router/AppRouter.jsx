import { Routes, Route, Navigate } from "react-router-dom"

import { CalendarPage } from "../calendar"
import { LoginPage } from "../auth"
import { useAuthStore } from "../hooks"
import { useEffect } from "react"

export const AppRouter = () => {
  // const authStatus = 'not-authenticated' // 'checking', 'authenticated', 'not-authenticated'
  const { status, checkAuthToken } = useAuthStore()

  useEffect(() => {
    checkAuthToken()
  }, [])

  if (status === 'checking') {
    return (
      <h3>Loading...</h3>
    )
  }


  return (
    <Routes>
      {
        (status === 'not-authenticated')
          ? <Route path="/auth/*" element={<LoginPage />} />
          : <Route path="/*" element={<CalendarPage />} />
      }

      <Route path="/*" element={<Navigate to='/auth/login' />} />
    </Routes>
  )
}