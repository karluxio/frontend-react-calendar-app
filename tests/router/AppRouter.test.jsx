import { render, screen } from "@testing-library/react";
import { AppRouter } from "../../src/router/AppRouter";
import { useAuthStore } from "../../src/hooks/useAuthStore";
import { MemoryRouter } from "react-router-dom";

jest.mock("../../src/hooks/useAuthStore");

jest.mock("../../src/calendar", () => ({
  CalendarPage: () => <h1>Mock Calendar Page</h1>,
}));

describe("<AppRouter />", () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  it("should shows loading and call checkAuthToken", () => {
    useAuthStore.mockReturnValue({
      status: "checking",
      checkAuthToken: mockCheckAuthToken,
    });

    render(<AppRouter />);
    // screen.debug();

    expect(screen.getByText("Loading..."));
    expect(mockCheckAuthToken).toHaveBeenCalled();
  });

  it("should shows login page if status is not authenticated", () => {
    useAuthStore.mockReturnValue({
      status: "not-authenticated",
      checkAuthToken: mockCheckAuthToken,
    });

    const { container } = render(
      <MemoryRouter initialEntries={["/any-route"]}>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText(/sign in/i));
    expect(screen.getByText(/register/i));
    expect(container).toMatchSnapshot();
  });

  it("should shows calendar page if status is authenticated", () => {
    useAuthStore.mockReturnValue({
      status: "authenticated",
      checkAuthToken: mockCheckAuthToken,
    });

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByText(/calendar/i));
  });
});
