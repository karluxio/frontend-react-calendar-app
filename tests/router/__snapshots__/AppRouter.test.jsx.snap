// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppRouter /> should shows login page if status is not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Sign In
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="loginPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Register
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="registerPassword2"
              placeholder="Repeat password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create an account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
